File: job_portal/src/app/core/services/communication.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { of } from 'rxjs';

export interface Announcement {
  id?: number;
  title: string;
  message: string;
  type: 'general' | 'maintenance' | 'feature' | 'urgent';
  target_audience: 'all' | 'job-seekers' | 'employers' | 'premium';
  send_methods: {
    email: boolean;
    in_app: boolean;
    push: boolean;
  };
  status: 'draft' | 'scheduled' | 'sent' | 'cancelled';
  scheduled_at?: string;
  sent_at?: string;
  created_by: number;
  created_by_name?: string;
  created_at: string;
  updated_at: string;
}

export interface Notification {
  id: number;
  user_id: number;
  title: string;
  message: string;
  type: 'announcement' | 'job_alert' | 'application_update' | 'system' | 'custom';
  announcement_id?: number;
  is_read: boolean;
  read_at?: string;
  priority: 'low' | 'medium' | 'high' | 'urgent';
  action_url?: string;
  metadata?: any;
  created_at: string;
  updated_at: string;
}

export interface NotificationPreferences {
  id?: number;
  user_id: number;
  email_announcements: boolean;
  email_job_alerts: boolean;
  email_application_updates: boolean;
  in_app_announcements: boolean;
  in_app_job_alerts: boolean;
  in_app_application_updates: boolean;
  push_announcements: boolean;
  push_job_alerts: boolean;
  push_application_updates: boolean;
}

export interface AnnouncementStats {
  total_announcements: number;
  sent_announcements: number;
  scheduled_announcements: number;
  draft_announcements: number;
  active_notifications: number;
}

@Injectable({
  providedIn: 'root'
})
export class CommunicationService {
  private baseUrl = 'http://localhost:3001/api/communication';
  private unreadCountSubject = new BehaviorSubject<number>(0);
  public unreadCount$ = this.unreadCountSubject.asObservable();

  constructor(private http: HttpClient) {
    this.loadUnreadCount();
  }

  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem('authToken');
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json'
    });
  }

  // Admin - Announcement Management
  createAnnouncement(announcement: Partial<Announcement>): Observable<any> {
    return this.http.post(`${this.baseUrl}/announcements`, announcement, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error creating announcement:', error);
        return of({ success: false, message: 'Failed to create announcement' });
      })
    );
  }

  getAnnouncements(page: number = 1, limit: number = 10): Observable<any> {
    return this.http.get(`${this.baseUrl}/announcements?page=${page}&limit=${limit}`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error fetching announcements:', error);
        return of({ success: false, announcements: [], total: 0 });
      })
    );
  }

  getAnnouncementStats(): Observable<any> {
    return this.http.get(`${this.baseUrl}/announcements/stats`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error fetching announcement stats:', error);
        return of({ 
          success: false, 
          stats: {
            total_announcements: 0,
            sent_announcements: 0,
            scheduled_announcements: 0,
            draft_announcements: 0,
            active_notifications: 0
          }
        });
      })
    );
  }

  // User - Notification Management
  getUserNotifications(page: number = 1, limit: number = 20): Observable<any> {
    return this.http.get(`${this.baseUrl}/notifications?page=${page}&limit=${limit}`, {
      headers: this.getAuthHeaders()
    }).pipe(
      map((response: any) => {
        if (response.success) {
          this.unreadCountSubject.next(response.unread);
        }
        return response;
      }),
      catchError(error => {
        console.error('Error fetching notifications:', error);
        return of({ success: false, notifications: [], total: 0, unread: 0 });
      })
    );
  }

  markNotificationAsRead(notificationId: number): Observable<any> {
    return this.http.put(`${this.baseUrl}/notifications/${notificationId}/read`, {}, {
      headers: this.getAuthHeaders()
    }).pipe(
      map((response: any) => {
        if (response.success) {
          this.loadUnreadCount();
        }
        return response;
      }),
      catchError(error => {
        console.error('Error marking notification as read:', error);
        return of({ success: false, message: 'Failed to mark notification as read' });
      })
    );
  }

  markAllNotificationsAsRead(): Observable<any> {
    return this.http.put(`${this.baseUrl}/notifications/read-all`, {}, {
      headers: this.getAuthHeaders()
    }).pipe(
      map((response: any) => {
        if (response.success) {
          this.unreadCountSubject.next(0);
        }
        return response;
      }),
      catchError(error => {
        console.error('Error marking all notifications as read:', error);
        return of({ success: false, message: 'Failed to mark all notifications as read' });
      })
    );
  }

  // Load unread count
  loadUnreadCount(): void {
    this.getUserNotifications(1, 1).subscribe(response => {
      if (response.success) {
        this.unreadCountSubject.next(response.unread);
      }
    });
  }

  // Notification Preferences
  getNotificationPreferences(): Observable<any> {
    return this.http.get(`${this.baseUrl}/preferences`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error fetching notification preferences:', error);
        return of({ success: false, preferences: null });
      })
    );
  }

  updateNotificationPreferences(preferences: Partial<NotificationPreferences>): Observable<any> {
    return this.http.put(`${this.baseUrl}/preferences`, preferences, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error updating notification preferences:', error);
        return of({ success: false, message: 'Failed to update preferences' });
      })
    );
  }

  // Admin - Send Custom Notification
  sendCustomNotification(notification: {
    user_id: number;
    title: string;
    message: string;
    type?: string;
    priority?: string;
    action_url?: string;
  }): Observable<any> {
    return this.http.post(`${this.baseUrl}/notifications/send`, notification, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error sending custom notification:', error);
        return of({ success: false, message: 'Failed to send notification' });
      })
    );
  }

  // Utility method to format notification time
  formatNotificationTime(timestamp: string): string {
    const now = new Date();
    const notificationTime = new Date(timestamp);
    const diffInMs = now.getTime() - notificationTime.getTime();
    const diffInSeconds = Math.floor(diffInMs / 1000);
    const diffInMinutes = Math.floor(diffInSeconds / 60);
    const diffInHours = Math.floor(diffInMinutes / 60);
    const diffInDays = Math.floor(diffInHours / 24);

    if (diffInSeconds < 60) {
      return 'Just now';
    } else if (diffInMinutes < 60) {
      return `${diffInMinutes}m ago`;
    } else if (diffInHours < 24) {
      return `${diffInHours}h ago`;
    } else if (diffInDays < 7) {
      return `${diffInDays}d ago`;
    } else {
      return notificationTime.toLocaleDateString();
    }
  }

  // Get scheduled announcements
  getScheduledAnnouncements(): Observable<any> {
    return this.http.get(`${this.baseUrl}/announcements/scheduled`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error fetching scheduled announcements:', error);
        return of({ success: false, message: 'Failed to fetch scheduled announcements', announcements: [] });
      })
    );
  }

  // Get draft announcements
  getDraftAnnouncements(page: number = 1, limit: number = 10): Observable<any> {
    return this.http.get(`${this.baseUrl}/announcements/drafts?page=${page}&limit=${limit}`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error fetching draft announcements:', error);
        return of({ success: false, announcements: [], total: 0 });
      })
    );
  }

  // Update announcement
  updateAnnouncement(id: number, announcement: Partial<Announcement>): Observable<any> {
    return this.http.put(`${this.baseUrl}/announcements/${id}`, announcement, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error updating announcement:', error);
        return of({ success: false, message: 'Failed to update announcement' });
      })
    );
  }

  // Cancel scheduled announcement
  cancelScheduledAnnouncement(announcementId: string): Observable<any> {
    return this.http.delete(`${this.baseUrl}/announcements/${announcementId}/cancel`, {
      headers: this.getAuthHeaders()
    }).pipe(
      catchError(error => {
        console.error('Error cancelling scheduled announcement:', error);
        return of({ success: false, message: 'Failed to cancel scheduled announcement' });
      })
    );
  }
}

File: job_portal_fastapi/app/services/communication_service.py
from sqlalchemy.orm import Session
from app.db.models.communication.announcement_model import Announcement
from app.db.models.communication.notification_model import Notification
from app.db.models.communication.notification_preference_model import NotificationPreference
from app.db.schemas.communication.announcement_schema import AnnouncementCreate, AnnouncementUpdate
from app.db.schemas.communication.notification_preference_schema import NotificationPreferenceUpdate
from fastapi import HTTPException
from datetime import datetime

# Announcements
def get_announcements(db: Session, status: str = None):
    query = db.query(Announcement)
    if status:
        query = query.filter(Announcement.status == status)
    return query.all()

def create_announcement(db: Session, announcement: AnnouncementCreate, user_id: str):
    db_announcement = Announcement(**announcement.dict(), created_by=user_id)
    db.add(db_announcement)
    db.commit()
    db.refresh(db_announcement)
    return db_announcement

def update_announcement(db: Session, announcement_id: str, announcement_update: AnnouncementUpdate):
    db_announcement = db.query(Announcement).filter(Announcement.id == announcement_id).first()
    if not db_announcement:
        raise HTTPException(status_code=404, detail="Announcement not found")
    update_data = announcement_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_announcement, key, value)
    db.add(db_announcement)
    db.commit()
    db.refresh(db_announcement)
    return db_announcement

def cancel_scheduled_announcement(db: Session, announcement_id: str):
    db_announcement = db.query(Announcement).filter(Announcement.id == announcement_id).first()
    if not db_announcement:
        raise HTTPException(status_code=404, detail="Announcement not found")
    if db_announcement.status != 'scheduled':
        raise HTTPException(status_code=400, detail="Announcement is not scheduled")
    db_announcement.status = 'cancelled'
    db.add(db_announcement)
    db.commit()
    db.refresh(db_announcement)
    return db_announcement

# Notifications
def get_user_notifications(db: Session, user_id: str):
    return db.query(Notification).filter(Notification.user_id == user_id).order_by(Notification.created_at.desc()).all()

def mark_notification_as_read(db: Session, notification_id: str, user_id: str):
    db_notification = db.query(Notification).filter(Notification.id == notification_id, Notification.user_id == user_id).first()
    if not db_notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    db_notification.is_read = True
    db_notification.read_at = datetime.utcnow()
    db.add(db_notification)
    db.commit()
    db.refresh(db_notification)
    return db_notification

def mark_all_notifications_as_read(db: Session, user_id: str):
    db.query(Notification).filter(Notification.user_id == user_id, Notification.is_read == False).update({"is_read": True, "read_at": datetime.utcnow()})
    db.commit()
    return {"message": "All notifications marked as read"}

# Notification Preferences
def get_notification_preferences(db: Session, user_id: str):
    preferences = db.query(NotificationPreference).filter(NotificationPreference.user_id == user_id).first()
    if not preferences:
        # Create default preferences if they don't exist
        preferences = NotificationPreference(user_id=user_id)
        db.add(preferences)
        db.commit()
        db.refresh(preferences)
    return preferences

def update_notification_preferences(db: Session, user_id: str, preferences_update: NotificationPreferenceUpdate):
    db_preferences = get_notification_preferences(db, user_id)
    update_data = preferences_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_preferences, key, value)
    db.add(db_preferences)
    db.commit()
    db.refresh(db_preferences)
    return db_preferences
